Improve SSE User Experience

VGEAREN opened the pull request:
<!-- Thank you for your contribution! -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

[Feature Enhancement] Improve SSE User Experience

This PR aims to enhance the user experience of SSE (Server-Sent Events). The following changes have been made:

  * Replace `fastapi.StreamingResponse` with `sse_starlette.sse.EventSourceResponse`
## Related issue number (if applicable)

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've run `format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed.
- [x] I've made sure the relevant tests are passing (if applicable).

alanxmay commented:
@VGEAREN I have made a similar modification before, but it has a problem that it is not compatible with the openai python sdk, because it will **send a ping event** every once in a while, which requires special handling by the client. Have you checked out this recently added feature? See [OpenAI-Compatible RESTful APIs](https://github.com/lm-sys/FastChat/blob/main/docs/openai_api.md).

VGEAREN commented:
> @VGEAREN I have made a similar modification before, but it has a problem that it is not compatible with the openai python sdk, because it will **send a ping event** every once in a while, which requires special handling by the client. Have you checked out this recently added feature? See [OpenAI-Compatible RESTful APIs](https://github.com/lm-sys/FastChat/blob/main/docs/openai_api.md).

yes, I think use a high ping interval can fix it.

merrymercy commented on fastchat/serve/openai_api_server.py:
```diff
@@ -90,11 +91,25 @@ async def check_model(request) -> Optional[JSONResponse]:
     return ret
 
-
+@app.post("/v1/chat/completions")
```
This is already a registration for this endpoint at https://github.com/lm-sys/FastChat/blob/f5c3f255578e2b0bf4ed252ea4bb4612a87c4d5c/fastchat/serve/openai_api_server.py#L287

Can we merge this modification to the existing endpoint?

VGEAREN replied:
solved

ojbk1984 reviewed (commented):
The code has been merged and conflicts resolved.

merrymercy commented:
@VGEAREN I am not very familiar with SSE. Could you educate me on how this is better than `StreamResponse`?
Also, did your current PR fully solve the issues mentioned by @alanxmay ?

VGEAREN commented:
> @VGEAREN I am not very familiar with SSE. Could you educate me on how this is better than `StreamingResponse `? Also, does your current PR fully solve the issues mentioned by @alanxmay ?


I integrated the OpenAI API and Fastchat API in a project(JAVA). After completing the OpenAI integration successfully, I attempted to use the same request method for Fastchat. However, I realized that it didn't work as expected. Eventually, I discovered that the issue was with the StreamingResponse. Therefore, I switched to EventSourceResponse, and it worked fine.

And yes, I have fixed the issue.

merrymercy commented:
@VGEAREN Thanks. It is merged.

merrymercy commented:
Sorry, I have to revert this PR because it introduces some errors I listed at #1875. Feel free to fix them and open a new PR if you still want this feature in the future version of FastChat.
